lib/salestation/rspec/glia_input_validation_error_matcher.rb

```diff
@@ -15,49 +15,68 @@ def initialize
         @field_error_messages = {}
       end
 
-      def on(field)
-        @fields << field
+      def on(*nested_fields)
+        @fields << nested_fields
         self
       end
 
       def with_type(*types)
-        @field_error_types[@fields.last] = types
+        @field_error_types[field_to_key(@fields.last)] = types
         self
       end
 
       def with_message(*messages)
-        @field_error_messages[@fields.last] = messages
+        @field_error_messages[field_to_key(@fields.last)] = messages
         self
       end
 
       def matches?(actual)
         @fields.all? do |field|
-          check_field_exists(field, actual) &&
+          check_field_exists(actual, *field) &&
             check_field_error_types(field, actual) &&
             check_field_error_messages(field, actual)
         end
       end
 
       private
 
-      def check_field_exists(field, actual)
-        actual[:error_details].key?(field)
+      def check_field_exists(actual, field, *nested_fields)
+        actual[:error_details].key?(field) &&
+          (
+            nested_fields.empty? ||
+            actual[:error_details][field].any? do |nested_hash|
+              check_field_exists(nested_hash, *nested_fields)
+            end
+          )
       end
 
       def check_field_error_types(field, actual)
-        return true unless @field_error_types[field]
+        return true unless @field_error_types[field_to_key(field)]
 
-        actual_error_types = actual[:error_details].fetch(field).map { |f| f.fetch(:type) }
-
-        (@field_error_types[field] - actual_error_types).empty?
+        actual_error_types = fetch_field_attribute_nested(actual, :type, *field)
+        (@field_error_types[field_to_key(field)] - actual_error_types).empty?
       end
 
       def check_field_error_messages(field, actual)
-        return true unless @field_error_messages[field]
+        return true unless @field_error_messages[field_to_key(field)]
+
+        actual_error_messages = fetch_field_attribute_nested(actual, :message, *field)
+        (@field_error_messages[field_to_key(field)] - actual_error_messages).empty?
+      end
 
-        actual_error_messages = actual[:error_details].fetch(field).map { |f| f.fetch(:message) }
 
-        (@field_error_messages[field] - actual_error_messages).empty?
+      def fetch_field_attribute_nested(actual, attribute, field, *nested_fields)
+        if nested_fields.empty?
+          actual[:error_details].fetch(field).map { |f| f.fetch(attribute) }
+        else
+          actual[:error_details][field].find do |error_hash|
+            return fetch_field_attribute_nested(error_hash, attribute, *nested_fields)
```

Comment:
`return` shouldn't be needed

Reply:
hm, actually no, it is needed. Could it be done without an early return?

Reply:
With variable assignment? I guess it not cleaner bit is it clearer?
```
     def fetch_field_attribute_nested(actual, attribute, field, *nested_fields)
        result = nil

        if nested_fields.empty?
          result = actual[:error_details].fetch(field).map { |f| f.fetch(attribute) }
        else
          actual[:error_details][field].find do |error_hash|
            result = fetch_field_attribute_nested(error_hash, attribute, *nested_fields)
          end
        end

        result
     end
```

Reply:
Usually, with `find` you expect it to return the first truthy result. Using `each` in the last example would be a bit better.

But maybe something like this:
```
actual[:error_details][field]
  .lazy
  .map {|error_hash| fetch_field_attribute_nested(error_hash, attribute, *nested_fields)}
  .reject(&:nil?)
  .first
```
I'll leave it up to you to decide which one is more readable.

Reply:
I could `break`

```
actual[:error_details][field].find do |error_hash|
  result = fetch_field_attribute_nested(error_hash, attribute, *nested_fields)
  break if result
end
```
Then it would still be first.
Currently working with hashes so it doesn't have other than semantic difference